feat: Enforce api keys

MCarlomagno opened the pull request:
# Summary
This PR introduces an API key permissions system with a new proc macro `require_permissions` that enforces  access control on all API endpoints.

> The API key from the API_KEY environment variable has full system access and bypasses all permission checks.

### Examples
```rust
// Global permission - requires ability to list all relayers
#[require_permissions(["relayers:read"])]
#[get("/relayers")]
async fn list_relayers(...) -> impl Responder { ... }

// Scoped permission - requires access to specific relayer
#[require_permissions([("relayers:read", "relayer_id")])]
#[get("/relayers/{relayer_id}")]
async fn get_relayer(relayer_id: web::Path<String>, ...) -> impl Responder { ... }

// Supports dotted paths for nested parameters
#[require_permissions([("transactions:read", "path.relayer_id")])]
#[get("/relayers/{relayer_id}/transactions/{transaction_id}")]
async fn get_transaction_by_id(path: web::Path<TransactionPath>, ..
```

### Scopes
Permissions use a structured `PermissionGrant` model with two scopes:

```rust
pub struct PermissionGrant {
    pub action: String,
    pub scope: PermissionScope,
}

pub enum PermissionScope {
    Global,           // Access to all resources
    Ids(Vec<String>), // Access to specific resource IDs
}
```

Examples

The permission system supports wildcards for flexible access control:
`*:*` - Super admin (full system access)
`relayers:*` - All actions on relayers resource
`transactions:*` - All actions on transactions resource

## Testing Process
1. Setup
Start the server with multiple relayers configured (e.g., `sepolia-example` and `solana-example`).
2. Create Scoped API Key
Create an API key with restricted permissions to only access `sepolia-example`:
```bash
curl -X POST "http://localhost:8080/api/v1/api-keys" \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer ADMIN_API_KEY" \
    -d '{
      "name": "Sepolia Read Only",
      "permissions": [
        {
          "action": "relayers:read",
          "scope": ["sepolia-example"]
        }
      ]
    }'
```
Response includes the API key value.
3. Test Authorized Access
Access the permitted relayer (should succeed):
```bash
curl -X GET http://localhost:8080/api/v1/relayers/sepolia-example \
      -H "Authorization: Bearer SCOPED_API_KEY_VALUE"
```
4. Test Unauthorized Access
Try to access a different relayer (should fail with 403 Forbidden):
```bash
curl -X GET http://localhost:8080/api/v1/relayers/solana-example \
      -H "Authorization: Bearer SCOPED_API_KEY_VALUE"
```
```bash
Expected response:
{
  "error": "Insufficient permissions. Required: relayers:read for resource 'solana-example'"
}
```
6. Test Global Permissions
Create an API key with global read access:
```bash
curl -X POST "http://localhost:8080/api/v1/api-keys" \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer ADMIN_API_KEY" \
    -d '{
      "name": "Global Read Only",
      "permissions": [
        {
          "action": "relayers:read",
          "scope": "global"
        }
      ]
    }'
```
This key can access any relayer but cannot perform write operations.

7. Test Wildcard Permissions
Create a super admin API key:
```
curl -X POST "http://localhost:8080/api/v1/api-keys" \
    -H "Content-Type: application/json" \
    -H "Authorization: Bearer ADMIN_API_KEY" \
    -d '{
      "name": "Super Admin",
      "permissions": [
        {
          "action": "*:*",
          "scope": "global"
        }
      ]
    }'
```
This key has full system access equivalent to the environment API key.

## Impact in `openzeppelin-relayer-sdk`

No impact :) the setup will work exactly as the setup using the default key:
```ts
const config = new Configuration({
  basePath: 'http://localhost:8080',
  accessToken: '<your-api-key-here>', // Can be env API key OR a created API key with permissions
});

const relayersApi = new RelayersApi(config);
```

## Checklist

- [ ] Add a reference to related issues in the PR description.
- [ ] Add unit tests if applicable.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Implemented API Key Permissions System enabling granular access control with global and resource-scoped permission types.
  * Added permission validation to API endpoints, restricting operations based on assigned permissions.
  * Enabled creation and management of API keys with specific permissions via REST API.

* **Documentation**
  * Added comprehensive API Key Permissions System documentation to README covering structure, formats, and examples.
  * Updated OpenAPI specification with per-endpoint permission requirements.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

zeljkoX commented on macros/src/tests.rs:
```diff
@@ -0,0 +1,68 @@
+    fn test_macro_detects_template_parameters() {
+        // This is a compile-time test - if the macro compiles, it works
+        let tokens = quote::quote! {
+            #[require_permissions(["relayers:get:{relayer_id}"])]
```
can it work with multiple permissions

For example required permissions are to get specific relayer or to be able to get any relayer?

MCarlomagno replied:
It works with multiple ones, but in this specific case [`relayers:get:{relayer_id}`](relayers:get:{relayer_id}`) should match both a specific or all relayers (using wildcards).
- If API key contains `relayers:get:abc-abc-abc-abc` permission, then it passes. 
- If API key contains `relayers:get:*` permission, then also passes.

I think there might be endpoints where we have more than one as you said, e.g. if you want to enforce both `signing:execute:{relayer_id}` + `transactions:execute:{relayer_id}` in the same endpoint:
```rs
#[require_permissions(["signing:execute:{relayer_id}", "transactions:execute:{relayer_id}"])]
```
should work

zeljkoX reviewed (commented):
Thanks for your work!

In general i like the idea of macro at route level.

In the future for some advanced cases we would need standard methods to invoke to check permissions logic but for most cases macros would do the job.

I think we should just see how permissions should be defined.

I have added my two cents in the comments. In general i think permissions should be simple and we can define scope outside of permission string and use in the logic to check. This would eliminate need to define multiple permissions, * and for specific id and would keep permission system simpler.

MCarlomagno commented on src/api/routes/plugin.rs:
```diff
@@ -5,22 +5,27 @@ use crate::{
 }
 
 /// Calls a plugin method.
+#[require_permissions(["plugins:execute:{plugin_id}"])]
```
Yes this was the most complicated part, the challenge there is that we don't really control the plugin interactions at "routes" level, Maybe doing some extra validation inside the `relayer-api` in `/plugins` works

MCarlomagno commented on macros/src/tests.rs:
```diff
@@ -0,0 +1,68 @@
+    fn test_macro_detects_template_parameters() {
+        // This is a compile-time test - if the macro compiles, it works
+        let tokens = quote::quote! {
+            #[require_permissions(["relayers:get:{relayer_id}"])]
```
Agree! that makes it simpler :)

MCarlomagno commented:
Hi @zeljkoX I simplified the implementation by supporting macro with simple permissions + scope as defined in the PR description.

Now the remaining challenge is how to propagate the API key permissions to plugins - when invoking plugins, we must pass the api key and somehow limit execution based on permissions, but I believe this is something we can add in the future, for now we can assume that plugin access represents being a super-user of relayers (adding a warning about this in documentation).

Also I have added an extension to openapi route docs to reflect the required API key permissions in each endpoint using `x-required-permissions` entry

zeljkoX commented on src/bootstrap/config_processor.rs:
```diff
@@ -38,8 +38,7 @@ where
     let api_key_model = ApiKeyRepoModel::new(
         "default".to_string(),
```
delete api key logic might check this string and prevent deletion.

Use case:
- redis mode is used
- api key got stored 
- user restarted instance and provided different value for API_KEY env var.
Current validation would allow deletion in this case as it compares env var.
Not sure how much valid is above case but anyway, it could produce side effects

zeljkoX commented on src/utils/auth.rs:
```diff
@@ -28,15 +32,155 @@ pub fn check_authorization_header(req: &ServiceRequest, expected_key: &SecretStr
+    // Extract and validate Authorization header
+    let token = extract_token_from_request(req)?;
+
+    // Check if this is the environment API key (full access bypass)
```
Potentially we should get from api key repo always to avoid some edge cases in the future. 

In memory mode repo is populated as well with env var value?

zeljkoX commented on src/utils/auth.rs:
```diff
@@ -28,15 +32,155 @@ pub fn check_authorization_header(req: &ServiceRequest, expected_key: &SecretStr
+    // Check permissions for the specific resource ID
+    for required_action in required_actions {
+        if !has_permission_grant_for_id(&api_key.permissions, required_action, resource_id) {
+            let required_list = required_actions.join(" AND ");
```
here as well

zeljkoX commented on src/models/api_key/request.rs:
```diff
@@ -1,8 +1,8 @@
 use serde::{Deserialize, Serialize};
 
 #[derive(Serialize, Deserialize)]
 pub struct ApiKeyRequest {
```
I think we need to narrow down api key cration logic to prevent use cases like:
- user with scoped api key permission + api key create permission can create an api key with higher permissions than it has and take over system.

I guess we can do something like: prevent api key creation that has higher permissions than api key used for creating a new api key.

zeljkoX replied:
Also maybe we should inject defaults permissions, like api key should be always able to delete itself.
